File: src/services/chat/SocketService.ts
import type { Message, SendMessageData } from '@/types';
import { getBackendBase, SOCKET_EVENTS, generateClientId } from '@/utils';
import { io, Socket } from 'socket.io-client';

export interface ISocketService {
  connect(token: string): Promise<void>;
  disconnect(): void;
  isConnected(): boolean;
  onMessage(callback: (message: Message) => void): void;
  onOnlineUsers(callback: (users: string[]) => void): void;
  onUnreadCounts(callback: (counts: { byPeer: Record<string, number>; total: number }) => void): void;
  onConnect(callback: () => void): void;
  onDisconnect(callback: (reason: string) => void): void;
  onConnectError(callback: (error: any) => void): void;
  sendMessage(data: SendMessageData): void;
  markAsRead(peer: string): void;
}

export class SocketService implements ISocketService {
  private socket: Socket | null = null;
  private token: string = '';
  private eventHandlers: Map<string, Set<Function>> = new Map();
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;

  constructor() {
    this.setupEventHandlers();
  }

  async connect(token: string): Promise<void> {
    if (this.socket?.connected) {
      return;
    }

    this.token = token;
    this.socket = io(getBackendBase(), {
      auth: { token },
      reconnection: true,
      reconnectionAttempts: Infinity,
      reconnectionDelay: 500,
      reconnectionDelayMax: 4000,
      timeout: 8000,
      transports: ['websocket', 'polling'],
    });

    this.setupSocketListeners();
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
  }

  isConnected(): boolean {
    return this.socket?.connected || false;
  }

  onMessage(callback: (message: Message) => void): void {
    this.addEventListener(SOCKET_EVENTS.PRIVATE_MESSAGE, callback);
  }

  onOnlineUsers(callback: (users: string[]) => void): void {
    this.addEventListener(SOCKET_EVENTS.ONLINE_USERS, callback);
  }

  onUnreadCounts(callback: (counts: { byPeer: Record<string, number>; total: number }) => void): void {
    this.addEventListener(SOCKET_EVENTS.UNREAD_COUNTS, callback);
  }

  onConnect(callback: () => void): void {
    this.addEventListener(SOCKET_EVENTS.CONNECTION, callback);
  }

  onDisconnect(callback: (reason: string) => void): void {
    this.addEventListener(SOCKET_EVENTS.DISCONNECT, callback);
  }

  onConnectError(callback: (error: any) => void): void {
    this.addEventListener(SOCKET_EVENTS.CONNECT_ERROR, callback);
  }

  sendMessage(data: SendMessageData): void {
    if (!this.socket?.connected) {
      console.error('Socket not connected');
      return;
    }

    const messageData = {
      ...data,
      clientId: data.clientId || generateClientId(),
    };

    this.socket.emit(SOCKET_EVENTS.PRIVATE_MESSAGE, messageData);
  }

  markAsRead(peer: string): void {
    if (!this.socket?.connected) {
      console.error('Socket not connected');
      return;
    }

    this.socket.emit(SOCKET_EVENTS.MARK_READ, { peer });
  }

  private setupEventHandlers(): void {
    this.eventHandlers.set(SOCKET_EVENTS.PRIVATE_MESSAGE, new Set());
    this.eventHandlers.set(SOCKET_EVENTS.ONLINE_USERS, new Set());
    this.eventHandlers.set(SOCKET_EVENTS.UNREAD_COUNTS, new Set());
    this.eventHandlers.set(SOCKET_EVENTS.CONNECTION, new Set());
    this.eventHandlers.set(SOCKET_EVENTS.DISCONNECT, new Set());
    this.eventHandlers.set(SOCKET_EVENTS.CONNECT_ERROR, new Set());
  }

  private setupSocketListeners(): void {
    if (!this.socket) return;

    this.socket.on(SOCKET_EVENTS.CONNECT, () => {
      console.log('Socket connected');
      this.reconnectAttempts = 0;
      this.emitEvent(SOCKET_EVENTS.CONNECTION);
    });

    this.socket.on(SOCKET_EVENTS.DISCONNECT, (reason: string) => {
      console.log('Socket disconnected:', reason);
      this.emitEvent(SOCKET_EVENTS.DISCONNECT, reason);
    });

    this.socket.on(SOCKET_EVENTS.CONNECT_ERROR, (error: any) => {
      console.error('Socket connection error:', error);
      this.reconnectAttempts++;
      this.emitEvent(SOCKET_EVENTS.CONNECT_ERROR, error);
    });

    this.socket.on(SOCKET_EVENTS.PRIVATE_MESSAGE, (message: Message) => {
      this.emitEvent(SOCKET_EVENTS.PRIVATE_MESSAGE, message);
    });

    this.socket.on(SOCKET_EVENTS.ONLINE_USERS, (users: string[]) => {
      this.emitEvent(SOCKET_EVENTS.ONLINE_USERS, users);
    });

    this.socket.on(SOCKET_EVENTS.UNREAD_COUNTS, (counts: { byPeer: Record<string, number>; total: number }) => {
      this.emitEvent(SOCKET_EVENTS.UNREAD_COUNTS, counts);
    });
  }

  private addEventListener(event: string, callback: Function): void {
    const handlers = this.eventHandlers.get(event);
    if (handlers) {
      handlers.add(callback);
    }
  }

  private emitEvent(event: string, ...args: any[]): void {
    const handlers = this.eventHandlers.get(event);
    if (handlers) {
      handlers.forEach(handler => {
        try {
          handler(...args);
        } catch (error) {
          console.error(`Error in ${event} handler:`, error);
        }
      });
    }
  }
}

File: flutter_im_client/lib/services/socket_service.dart
import 'dart:async';
import 'package:socket_io_client/socket_io_client.dart' as io;
import 'package:logging/logging.dart';
import '../constants/api_constants.dart';
import '../models/message.dart';
import '../utils/exceptions.dart';
import 'auth_service.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  io.Socket? _socket;
  final AuthService _authService = AuthService();
  final Logger _logger = Logger('SocketService');

  final StreamController<Message> _messageController = StreamController<Message>.broadcast();
  final StreamController<List<String>> _onlineUsersController = StreamController<List<String>>.broadcast();
  final StreamController<Map<String, dynamic>> _unreadCountsController = StreamController<Map<String, dynamic>>.broadcast();

  Stream<Message> get onMessageReceived => _messageController.stream;
  Stream<List<String>> get onOnlineUsersUpdated => _onlineUsersController.stream;
  Stream<Map<String, dynamic>> get onUnreadCountsUpdated => _unreadCountsController.stream;

  bool get isConnected => _socket?.connected ?? false;

  Future<void> connect() async {
    if (_socket?.connected == true) {
      _logger.info('Socket already connected');
      return;
    }

    final token = await _authService.getAccessToken();
    if (token == null) {
      throw Exception('No authentication token available');
    }

    _socket = io.io(
      ApiConstants.socketUrl,
      {
        'transports': ['websocket'],
        'autoConnect': true,
        'reconnection': true,
        'auth': {'token': token},
      },
    );

    _setupEventListeners();
    
    _socket!.connect();
    _logger.info('Socket connection initiated');
  }

  void _setupEventListeners() {
    _socket!.onConnect((_) {
      _logger.info('Socket connected');
    });

    _socket!.onDisconnect((_) {
      _logger.info('Socket disconnected');
    });

    _socket!.onConnectError((error) {
      _logger.severe('Socket connection error: $error');
    });

    _socket!.onError((error) {
      _logger.severe('Socket error: $error');
    });

    // Listen for private messages
    _socket!.on(ApiConstants.privateMessage, (data) async {
      try {
        final me = await _authService.getCurrentUser();
        final m = Map<String, dynamic>.from(data as Map);
        final from = (m['from'] ?? '') as String;
        final to = (m['to'] ?? '') as String;
        final ts = (m['timestamp'] ?? DateTime.now().millisecondsSinceEpoch) as int;
        final id = '${ts}_${from}_${to}_${(m['clientId'] ?? '')}';
        final message = Message(
          id: id,
          from: from,
          to: to,
          content: (m['content'] ?? '') as String,
          timestamp: DateTime.fromMillisecondsSinceEpoch(ts),
          clientId: m['clientId'] as String?,
          isOwn: from == me.id,
        );
        _messageController.add(message);
        _logger.info('Received private message from ${message.from}');
      } catch (e) {
        _logger.severe('Error parsing private message: $e');
      }
    });

    // Listen for online users update
    _socket!.on(ApiConstants.onlineUsers, (data) {
      try {
        final onlineUsers = List<String>.from(data as List);
        _onlineUsersController.add(onlineUsers);
        _logger.info('Online users updated: ${onlineUsers.length} users');
      } catch (e) {
        _logger.severe('Error parsing online users: $e');
      }
    });

    // Listen for unread counts update
    _socket!.on(ApiConstants.unreadCounts, (data) {
      try {
        final unreadCounts = Map<String, dynamic>.from(data as Map);
        _unreadCountsController.add(unreadCounts);
        _logger.info('Unread counts updated');
      } catch (e) {
        _logger.severe('Error parsing unread counts: $e');
      }
    });
  }

  void sendMessage(String to, String content, {String? clientId}) {
    if (!isConnected) {
      throw ChatException(
        message: 'Socket not connected',
        type: ChatErrorType.connectionFailed,
      );
    }

    final messageData = {
      'to': to,
      'content': content,
      'clientId': clientId,
    };

    _socket!.emit(ApiConstants.privateMessage, messageData);
    _logger.info('Sent private message to $to');
  }

  void markAsRead(String peer) {
    if (!isConnected) {
      _logger.warning('Cannot mark as read: Socket not connected');
      return;
    }

    _socket!.emit(ApiConstants.markRead, {'peer': peer});
    _logger.info('Marked conversation as read: $peer');
  }

  Future<void> disconnect() async {
    if (_socket != null) {
      _socket!.disconnect();
      _socket!.dispose();
      _socket = null;
      _logger.info('Socket disconnected and disposed');
    }
  }

  void dispose() {
    disconnect();
    _messageController.close();
    _onlineUsersController.close();
    _unreadCountsController.close();
  }
}
